native/src/tablecloth.ml

```diff
@@ -6,6 +6,133 @@ let ( << ) (f1 : 'b -> 'c) (f2 : 'a -> 'b) : 'a -> 'c = fun x -> x |> f2 |> f1
+
+  let to_list = toList
+
+  let toIndexedList xs = 
+    Base.Array.fold_right xs ~init:(length xs - 1, []) ~f:(fun x (i, acc) -> 
+      (i - 1, ((i, x) :: acc)))
+    |> Base.snd
+  
+  let to_indexed_list = toIndexedList
+
+  let get ~index a = 
+    if index >= 0 && index < length a then Some (Base.Array.get a index) else None
+
+  let set ~index ~value a = Base.Array.set a index value
+
+  let sum (a : int array) : int = Base.Array.fold a ~init:0 ~f:( + )
+
+  let floatSum (a : float array) : float = Base.Array.fold a ~init:0.0 ~f:( +. )
+
+  let float_sum = floatSum
+
+  let filter ~(f : 'a -> bool) (a : 'a array) : 'a array = Base.Array.filter a ~f
+
+  let map ~(f : 'a -> 'b) (a : 'a array) : 'b array = Base.Array.map a ~f
+
+  let mapWithIndex  ~(f : 'int -> 'a -> 'b) (a : 'a array) : 'b array = Base.Array.mapi a ~f
+  
+  let map_with_index = mapWithIndex
+  
+  let mapi = mapWithIndex 
+
+  let map2 ~(f : 'a -> 'b -> 'c) (a : 'a array) (b : 'b array) : 'c array =
+    let minLength = min (length a) (length b) in
+    Base.Array.init minLength ~f:(fun i -> f a.(i) b.(i))
+
+  let map3 ~(f : 'a -> 'b -> 'c -> 'd) (arrayA : 'a array) (arrayB : 'b array) (arrayC : 'c array) : 'd array =
+    let minLength : int = Base.Array.fold ~f:Base.min ~init:(length arrayA) [|length arrayB; length arrayC|] in
+    Base.Array.init minLength ~f:(fun i -> f arrayA.(i) arrayB.(i) arrayC.(i))
+
+  let flatMap ~f a = Base.Array.concat_map a ~f
+
+  let flat_map = flatMap
+
+  let find ~(f : 'a -> bool) (a : 'a array) : 'a option = Base.Array.find a ~f
+
+  let append (a : 'a array) (a' : 'a array)  : 'a array = Base.Array.append a a'
+
+  let concatenate  (al : 'a array array) : 'a array = Base.Array.concat (Base.Array.to_list al)
+
+  let flatten = concatenate
+
+  let intersperse ~sep array = 
+    Base.Array.init (max 0 (Array.length array * 2 - 1)) ~f:(fun i -> 
+      if i mod 2 <> 0 then sep else array.(i / 2)
+    )
+
+  let any ~(f : 'a -> bool) (a : 'a array) : bool = Base.Array.exists ~f a
+
+  let all ~(f : 'a -> bool) (a : 'a array) : bool = Base.Array.for_all ~f a
+
+  let slice ~from ?to_ array =
+    let defaultTo = match to_ with 
+      | None -> length array
+      | Some i -> i
+    in
+    let sliceFrom = 
+      if from >= 0 then min (length array) from 
+      else max 0 (min (length array) (length array + from))
+    in    
+    let sliceTo = 
+      if defaultTo >= 0 then min (length array) defaultTo 
+      else max 0 (min (length array) (length array + defaultTo))
+    in    
+    
+    if sliceFrom >= sliceTo then empty else (
+      Array.init (sliceTo - sliceFrom) (fun i -> array.(i + sliceFrom))
+    )
+
+  let foldLeft ~(f : 'a -> 'b -> 'b) ~(initial : 'b) (a : 'a array) : 'b =
+    Base.Array.fold ~f:(fun b a -> f a b) ~init:initial a
+
+  let fold_left = foldLeft
+  
+  let foldRight ~(f : 'a -> 'b -> 'b) ~(initial : 'b) (a : 'a array) : 'b =
+    Base.Array.fold_right ~f ~init:initial a
+
+  let fold_right = foldRight
+
```

Comment:
alias to foldr?

Reply:
Its a slightly different signature (initial rather than init).

Reply:
```suggestion
let foldr = foldRight
```

Reply:
Oh you mean add the `foldr` alias in addition to `foldRight` and `fold_right`.

This was an intentional deviation from Elms standard library as I felt it was more in keeping with Elms guidelines on API's: https://package.elm-lang.org/help/design-guidelines#use-human-readable-names

In this particular case the Elm docs even go on to explain that `foldr` means "fold from the right".

Reply:
im in a rush, gonna read it when i have time. My thought was we have indexedMap/mapi

Reply:
I think we should remove mapi and all synonyms - have just one way to do it.